### ISSUE TITLE
PDF/A-3 support

### ISSUE THREAD

**Issue by xsuchy09** [qpdf/qpdf#935]
There is missing support for PDF/A-3. Required attribute AFRelationship of attached file is missing.

> 3.2 PDF/A-3’s restrictions
> PDF/A-3, which introduced Associated Files to PDF in 2012, imposes some additional
> requirements that are not present in PDF 2.0:
> • PDF/A-3 excludes external file references, so associations can be specified only for
> embedded files
> • Since PDF/A-3 requires the AF entry for all embedded files, each file’s MIME-type is
> always required
> • PDF/A-3 requires every embedded file to be an Associated File and thus requires
> the AFRelationship entry in each case
> • The list of objects for which the AF entry may be present is explicitly open in PDF
> 2.0 and an AF entry may be used anywhere in a PDF file. However, in PDF/A-3 this
> list is closed and the AF entry may only be present in the locations explicitly
> identified in the standard.
> • PDF/A-3 has - similar to the list defined in PDF 2.0 - a list of possible objects in
> which the AF entry may be present (ISO 19005-3, Table E.1). However, the list in
> PDF/A-3 lacks the DPart entry

**Comment by jberkenbilt** [comment]
There is no expectation of qpdf supporting PDF/A-3. It is the intention that qpdf preserve PDF/A on a file that already supports it. Is there a situation in which qpdf is breaking PDF/A-3 on a file that already has it?

**Comment by xsuchy09** [comment]
Yes when you will try to add an attachment to the PDF/A-3 file. PDF is invalid after this operation.

**Comment by jberkenbilt** [comment]
This is not likely to be fixed anytime soon. When I say qpdf preserves PDF/A, I mean that it preserves it when just rewriting files, not when mutating them with operations like adding attachments. Basically qpdf has no PDF/A support at this time. I would hope that if you added an attachment with a tool that had PDF/A awareness and then transformed the file with qpdf, the file would remain in compliance.

**Comment by warpspin** [comment]
I, too, wish it were possible to add PDF/A-3 compliant attachments with qpdf. I can currently use the qpdf cli for most of my PDF manipulation tasks except this one thing where it seems (I'm not familiar with the format, so of course, it might in reality be supercomplicated) the only thing missing is being able to set the AFRelationship tag properly.

**Comment by jberkenbilt** [comment]
If someone would like to investigate what is required and it turns out to be simple, it is more likely that it would get squeezed in.

**Comment by warpspin** [comment]
Does this document help? https://pdfa.org/wp-content/uploads/2018/10/PDF20_AN002-AF.pdf

3.1 General
In PDF 2.0, Associated Files require:
• at least one AF entry in a PDF object (which can be the catalog if the AF entry
relates to the whole PDF)
• an AFRelationship entry in the file specification dictionary specifying the nature of
the relationship between the PDF (or a PDF object), and the related content. This
entry is required to contain one of the values defined in PDF 2.0: Source, Data,
Alternative, Supplement, EncryptedPayload, FormData, Schema or Unspecified.
Custom values may be used where none of these entries is appropriate.
For embedded Associated Files, the following requirements also apply:
• a MIME type (as defined in RFC 8118) must be specified in the Subtype entry of the
embedded file stream dictionary
• if the (recommended) Params entry is present, it shall specify the latest
modification date of the embedded file.
Referenced Associated Files are permitted, but in almost all contexts, only embedded files
make sense.
3.2 PDF/A-3’s restrictions
PDF/A-3, which introduced Associated Files to PDF in 2012, imposes some additional
requirements that are not present in PDF 2.0:
• PDF/A-3 excludes external file references, so associations can be specified only for
embedded files
• Since PDF/A-3 requires the AF entry for all embedded files, each file’s MIME-type is
always required
• PDF/A-3 requires every embedded file to be an Associated File and thus requires
the AFRelationship entry in each case
• The list of objects for which the AF entry may be present is explicitly open in PDF
2.0 and an AF entry may be used anywhere in a PDF file. However, in PDF/A-3 this
list is closed and the AF entry may only be present in the locations explicitly
identified in the standard.
• PDF/A-3 has - similar to the list defined in PDF 2.0 - a list of possible objects in
which the AF entry may be present (ISO 19005-3, Table E.1). However, the list in
PDF/A-3 lacks the DPart entry (since DPart was only defined in PDF/VT when
PDF/A-3 was developed).

**Comment by jberkenbilt** [comment]
@warpspin Sorry for the delayed response. This is very helpful. It doesn't look trivial though, so it will have to stay on the backlog until someone has time to work it. Thanks for this though -- it is a big help.

**Comment by hrvoj3e** [comment]
How about just adding `AFRelationship` when adding attachments? Maybe as an opt-in feature ?

**Comment by jberkenbilt** [comment]
@hrvoj3e Worth considering. No one is available to work on this right now. A pull request may be entertained.
